### ISSUE TITLE
autotune: cached matmul config fails validation under fusion

### ISSUE THREAD

**Issue by RAV64** [tracel-ai/burn#5352]
**Describe the bug**

With `fusion` + `autotune` enabled on CUDA, an autotuned matmul winner that benchmarked **Ok** later fails launch validation on a cache hit:

```
Should run when selected by autotune.: matmul_specialized_cyclic_mma: An unknown error happened.
Unable to launch matmul because the config is invalid: "Async copy requires strides to be aligned to 16 bytes"
```

The panic fires inside a fire-and-forget task on the device-runner thread (`DSU-0-0`), where it is caught and only **warn-logged** (`cubecl-common/src/device/handle/channel.rs`, the `Task` handlers `catch_unwind` + `log::warn!`). The op's output tensors were registered before execution, so after the swallow the fusion handle state is inconsistent: subsequent tasks panic with `Should have handle for tensor TensorId { .. }` (`burn-ir/src/handle.rs:97`) and `index out of bounds` in `burn-fusion/src/stream/execution/ordering.rs:73`, until a blocking readback propagates `CallError` (`burn-fusion/src/client.rs:200`).

On **0.22.0-pre.1** this same swallow was fully silent in our training: the learner kept stepping against never-written buffers and weights went NaN mid-run (loss → NaN at steps 241–771 across five runs; six multi-hour self-play campaigns ruined before we root-caused it). On **0.22.0-pre.2** the corruption is loud but still fatal. Disabling `fusion` eliminates everything (our production workloads run fusion-off since; ~8% throughput cost).

**To Reproduce**

~60 lines, deterministic (3/3 fresh-cache runs) on RTX 4080 SUPER, CUDA 12.9, burn 0.22.0-pre.2: step 0 trains, the first forward after the AdamW update dies.

Two one-line changes make it complete cleanly, which points at the mechanism:
- head width **7 → 8** passes (`anchor(7) = 8` in the autotune key);
- removing the `fusion` feature passes.

Batch 384 and 512 both fail; an all-128 net (input 126 → 128) still fails — the 7-wide head is the poisoned dimension.

*Cargo.toml*

```toml
[package]
name = "burn-fusion-repro"
version = "0.1.0"
edition = "2021"

[dependencies]
burn = { version = "=0.22.0-pre.2", default-features = false, features = [
  "std", "autodiff", "optim", "cuda", "fusion", "autotune",
] }
```

*src/main.rs*

```rust
use burn::module::Module;
use burn::nn::{Linear, LinearConfig};
use burn::optim::{AdamWConfig, GradientsParams};
use burn::tensor::activation::relu;
use burn::tensor::{Device, Distribution, Tensor};

#[derive(Module, Debug)]
struct Net {
    input: Linear,
    fc1: Linear,
    fc2: Linear,
    head: Linear,
}

impl Net {
    fn new(device: &Device) -> Self {
        let h = 128;
        Self {
            input: LinearConfig::new(126, h).init(device),
            fc1: LinearConfig::new(h, h).init(device),
            fc2: LinearConfig::new(h, h).init(device),
            head: LinearConfig::new(h, 7).init(device), // 7 -> 8 makes it pass
        }
    }

    fn forward(&self, x: Tensor<2>) -> Tensor<2> {
        let h = relu(self.input.forward(x));
        self.head.forward(self.fc2.forward(relu(self.fc1.forward(h))))
    }
}

fn main() {
    let device = Device::cuda(0).autodiff();
    let mut net = Net::new(&device);
    let mut optim = AdamWConfig::new().init();

    for step in 0..5 {
        let x = Tensor::<2>::random([512, 126], Distribution::Uniform(0.0, 1.0), &device);
        let loss = net.forward(x).powf_scalar(2.0).mean();
        let scalar = loss.clone().into_scalar::<f32>();
        assert!(scalar.is_finite(), "step {step}: NON-FINITE LOSS {scalar}");
        println!("step {step}: loss {scalar:.5}");
        let grads = GradientsParams::from_grads(loss.backward(), &net);
        net = optim.step(3e-4, net, grads);
    }
    println!("completed without corruption");
}
```

Note: the autotune cache persists under `./target/environment/` — delete it between runs for a clean reproduction.

Observed output:

```
step 0: loss 0.00419

thread 'DSU-0-0' panicked at cubecl-runtime-0.11.0-pre.2/src/tune/local.rs:161:18:
Should run when selected by autotune.: matmul_specialized_cyclic_mma: An unknown error happened.
Unable to launch matmul because the config is invalid: "Async copy requires strides to be aligned to 16 bytes"

thread 'main' panicked at burn-fusion-0.22.0-pre.2/src/client.rs:200:14:
called `Result::unwrap()` on an `Err` value: CallError(task panicked on device runner thread: ...)
```

In larger nets (add LayerNorm blocks / a second head) the same first panic is followed by the `Should have handle for tensor` / `ordering.rs` cascade before the `CallError` surfaces, and each intermediate panic is visible only as a `log::warn!`.

**Expected behavior**

The loop prints five finite, decreasing losses and `completed without corruption`. When a cached autotune winner fails launch validation, the tuner should fall back to another candidate (or re-tune) rather than panic; and a failed op on the device-runner thread should fail the stream loudly rather than leave registered-but-never-written tensor handles behind.

**Additional context**

*The failing problem* (instrumented the validator to print it): the forward head matmul `[512,128] @ [128,7]`, rhs strides `[7,1]` → canonical stride 28 bytes, under the 16-byte async-copy threshold checked by `validate_async_copy_with_problem` (`cubek-matmul/src/components/global/read/strategy/base.rs`).

*Defect 1 — the autotune key overstates stride alignment.* `MatmulAutotuneKey` computes `lhs_stride_factor` / `rhs_stride_factor` from the **anchored** dims, not the real strides (`cubek-matmul/src/strategy/tune_key.rs`: `stride_factor(k_anchored, …)`). `anchor(7) = 8` keys this problem as 32-byte aligned; `anchor(126) = 128` keys a 504-byte-stride problem as 512-byte aligned. Problems that fail the async-copy/TMA validators therefore share buckets with problems that pass them, and a winner tuned on an aligned member is an invalid config for the others — validated only *after* autotune has committed. (Our real workload's autotune cache shows these collision buckets directly: the `Linear(126, _)` forward/backward problems share buckets with the 128-wide hidden-layer ones.)

*Defect 2 — nothing gates these strategies on alignment, so an invalid config can win autotune.* `Strategy::SpecializedCyclicCmma` / `SpecializedCyclicMma` read through `AsyncPartialCyclicLoading`, but they are registered in the `accelerated` group of `burn-cubecl/src/kernel/matmul/tune/base.rs` with no alignment gate — while the `tma` group applies exactly the `stride_factor >= 4` threshold these need. An under-aligned problem therefore hands them `PRIORITY_MAX`, and the failure only appears at setup, after autotune has committed to the winner, where `.expect("Should run when selected by autotune.")` in `cubecl-runtime/src/tune/local.rs` turns it into a panic. The fix belongs in selection, not in recovery — see the comment below for the full root cause.

*Defect 3 — the panic is swallowed and the stream keeps going.* Fire-and-forget tasks on the device-runner thread run under `catch_unwind` and only `log::warn!` the payload (`cubecl-common/src/device/handle/channel.rs`). Outputs are registered before execution, so a panicked op leaves live, never-written tensor handles behind — this is the silent-NaN mechanism, and it is unchanged in pre.2 (only blocking calls propagate errors, via the oneshot `CallError` path). A failed task should poison the stream/queue loudly.

PRs: defect 1 (key reports real stride alignment) — https://github.com/tracel-ai/cubek/pull/488; defect 2 (gate the async-copy strategies on that alignment) — see the comment below. They are only meaningful together: the gate reads the stride factors off the key, so it cannot fire until the key stops deriving them from the anchored dims.

An earlier attempt at defect 2 made the tuner fall back to the remaining candidates on a launch failure (cubecl#1510). That was rejected upstream, correctly: it would have to hand the real inputs to a second candidate, and autotune must never clone them or in-place operations write into the copy. Preventing the selection is the right layer.

One more observation from validating those PRs against burn main's pinned revs (cubecl `f0721ea`, cubek `611491b`): on sm_89 this repro no longer crashes there — but only because the offending kernel now fails *compilation* at tune time with `UnsupportedOp("plane.elect")`, surfacing as another swallowed device-thread panic (defect 3), so it never becomes the cached winner. The repro's crash should be verified against the published `0.22.0-pre.2` crates as specified above; the defective `.expect`/swallow paths are unchanged on main.

Environment:

- burn 0.22.0-pre.2 (cubek-matmul 0.3.0-pre.2, cubecl 0.11.0-pre.2); pre.1 exhibits the silent-NaN variant
- RTX 4080 SUPER (sm_89), CUDA 12.9, Linux
- f32 (not bf16-specific)

**Comment by RAV64** [comment]
Root cause found, and it is narrower than this issue originally described. I have edited the body above to match; here is what changed and why.

**cubecl#1510 was the wrong layer.** @nathanielsimard closed it because a fallback would have to hand the real inputs to a second candidate, and autotune must never clone them — in-place operations would write into the copy. That is right, and the constraint pointed straight at where the actual bug is.

**The bug is a missing gate in the unfused matmul tuner.** `Strategy::SpecializedCyclicCmma` and `SpecializedCyclicMma` bind `L = AsyncPartialCyclicLoading` in `SpecializedAlgorithm<L, AL>`, and that single `L` serves both operands' partial loads. Its validation calls `validate_async_copy_with_problem`, which checks the real strides for 16-byte alignment at setup.

`burn-cubecl/src/kernel/matmul/tune/base.rs` already has that gate — the `tma` group returns `PRIORITY_NEVER` unless both stride factors clear the threshold. But it is attached only to the `*Tma*` strategies. The two specialized cyclic ones are in `accelerated` with no extra group, so nothing checks alignment for them and an under-aligned problem hands them `PRIORITY_MAX`. Every other entry in that list reads through sync loaders (`Simple*`, `DoubleCyclic*` → `CyclicDoubleBufferingAlgorithm`, `OrderedDouble*`), so they need no gate.

**One correction to the `fusion` label:** the defective code is in the *unfused* tuner. The fused tuner names its tunables `double_buffering_specialized_*` (not the `matmul_specialized_cyclic_mma` in the traces above) and maps them to `CyclicDoubleBufferingAlgorithm`, which is sync. Fusion is what presents the under-aligned view that the ungated strategy then cannot load, not where the defect lives — worth knowing, since the unfused path is exposed too.

**This also explains the contradiction in the original report.** Fixing the key alone did not stop the crash because splitting the autotune buckets does not help when nothing consults alignment for these strategies in the first place. Conversely, gating alone does nothing while the key still derives its stride factors from the anchored dims — `anchor(7) = 8` reports a 28-byte stride as 32-byte aligned. cubek#488 and the gate are a matched pair; either one merged alone is a no-op for this bug.

**Verified on a real CUDA training workload** (RTX 4080 SUPER, sm_89), with both fixes backported onto the 0.22.0-pre.2 releases: the gate fires on 14 of 20 matmul autotune keys — those two strategies are recorded with no benchmark result on them and tune normally on the other 6, so aligned shapes still see the full candidate pool. The training numerics suites pass on both the `cuda` and `cuda-fusion` configurations; previously the fused one died on the step after the first optimizer update.

**Defect 3 is unchanged and still open.** A task that panics on the device-runner thread is caught and only `log::warn!`-ed while its outputs stay registered and unwritten. This work removes the trigger we hit, but any other panic on that path still corrupts state silently, and that is what turned a launch error into six ruined training runs rather than a crash.

---

The gate is now open as https://github.com/tracel-ai/burn/pull/5370, paired with https://github.com/tracel-ai/cubek/pull/488.
